Tests for cluster features

piodul opened the pull request:
This PR adds a set of tests for cluster features. The set covers those tests from the test plan for the upcoming "cluster features on raft" that do not depend on implementation-specific details. Those tests are applicable to the existing gossip-based implementation, so they can be useful for us right now.

The tests simulate cluster upgrades by conditionally marking support for the TEST_ONLY_FEATURE on the nodes via error injection. Therefore, the tests work only in non-release mode.

The `test_partial_upgrade_can_be_finished_with_removenode` test is marked as skipped because of a bug in gossip that prevents features from being enabled if a node was removed within the last 3 days (#14194).

michoecho commented on test/pylib/util.py:
```diff
@@ -121,4 +121,28 @@ async def read_barrier(cql: Session, host: Host):
+# Wait for the given feature to be enabled.
+async def wait_for_feature(feature: str, cql: Session, host: Host, deadline: float) -> None:
+    async def feature_is_enabled():
+        rs = await cql.run_async("select value from system.scylla_local where key = 'enabled_features'", host=host)
+        if rs:
+            value = rs[0].value
+            if feature in value:
+                return True
+        return None
```
Note: could be shortened to:
```
return feature in get_enabled_features(cql, host) or None
```

piodul replied:
Good point, done

michoecho commented on test/topology/test_cluster_features.py:
```diff
@@ -0,0 +1,200 @@
+
+@pytest.mark.asyncio
+async def test_downgrade_after_partial_upgrade(manager: ManagerClient) -> None:
+    """Simulates an partial of a cluster by enabling the test features in all
```
"an partial" -> "a partial upgrade"
?

piodul replied:
Fixed, thanks

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1767/

piodul commented:
v2:
- Moved `get_supported_features`/`get_enabled_features` to a separate commit
- Reimplemented some parts of `wait_for_feature` via `get_enabled_features`
- Fixed a typo in a comment

piodul commented:
> CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1774/

It looks like sanity tests failed due to a spot termination. I'll rerun.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1776/

nyh commented on test/pylib/util.py:
```diff
@@ -124,13 +124,21 @@ async def read_barrier(cql: Session, host: Host):
+async def get_enabled_features(cql: Session, host: Host) -> set[str]:
+    """Returns a set of cluster features that a node considers to be enabled."""
+    rs = await cql.run_async(f"SELECT value FROM system.scylla_local WHERE key = 'enabled_features'", host=host)
+    return set(rs[0].value.split(","))
```
The code before this patch didn't seem to have the split()... Was it simply wrong?

michoecho replied:
If I understand you correctly, you mean the fact that before it was `string1 in string2` and now it's `string1 in string2.split(",")`?

The code before the patch was wrong, but it was OK as long as no feature name was a substring of another feature name. (Note that in Python `string1 in string2` checks whether `string1` is a substring of `string1`. So before the patch, we would check whether the feature name is a substring of the comma-separated string of features. After the patch, we explicitly construct the list (Python list) of features from the comma-separated string, and check whether the feature is in the list.

nyh replied:
Right, I forgot that "string in string" also "sort of, sometimes" works. Of course the new version with "string in set" is better.

piodul replied:
It used to work, but it was less explicit. `str1 in str2` in python is equivalent to "str1 is a substring of str2", so previously the function `wait_for_feature` checked whether the feature name is a _substring_ of the comma-separated string of features.

piodul replied:
> The code before the patch was wrong, but it was OK as long as no feature name was a substring of another feature name.

Right, I missed that possibility. So the previous version wasn't entirely correct after all.
